src/PlayerControls.cpp

```diff
@@ -6,25 +6,23 @@ void PlayerControlsSystem::on_tick()
 
     for (auto&& [player_entity, player, bindings, pos, sprite, flip] : access_storage().each())
     {
-        if (keys.is_pressed(bindings.up) && player.is_grounded) // keys.is_down() causes a bug where player can levitate trough platforms 
+        if (keys.is_pressed(bindings.up) && player.fsm.currentState != PlayerState::AIRBORNE) // keys.is_down() causes a bug where player can levitate trough platforms 
         {
             player.vertical_velocity = -JUMP_MOD;
-            player.is_grounded = false;
+            player.fsm.change_state(PlayerTransition::JUMP);
```

Comment:
Looking at this use, I'd recommend you change the name of `change_state` to `trigger` or `fire`. `change_state` suggests that a change of state will happen, and what we want is something that will trigger a change if possible, or do nothing.